### PR TITLE
KNOX-3426 - Add DelegationPolicyService schema, interface, and JDBC implementation

### PR DESCRIPTION
**Pull request by hsheinblatt**
KNOX-3426 - Add DelegationPolicyService schema, interface, and JDBC implementation

## What changes were proposed in this pull request?

This pull request adds the persistence layer for RFC 8693 delegation policies in Knox IDF.
Delegation policies control which actors (external token issuers and service accounts) are
authorized to perform cross-subject token exchange on behalf of specific users or groups
 and for specific resources and scopes.

A five-table schema is added to store policies: a core DELEGATION_REGISTRY table keyed on
(actor_authority, actor_id) with an allow_headless_exchange flag and an optional per-policy
token TTL; four child tables for per-user, per-group, per-resource, and per-resource-scope
entries. Scopes are defined per resource rather than as a flat list, so a policy can allow
different scopes for different target resources. A resource with no scope entries means all
scopes are allowed for that resource.

The DelegationPolicyService gateway service interface is added to gateway-spi, where it is
visible to the filter layer (gateway-provider-security-jwt) that will enforce policies during
token exchange. The interface includes CRUD operations for policy lifecycle management and an
evaluate() method that accepts a PolicyCheckRequest and returns a PolicyDecision.
PolicyCheckRequest and PolicyDecision are also added to gateway-spi as the inputs and outputs
of evaluate(). DelegationPolicy, the POJO representing a stored policy, is also in gateway-spi
since it is returned by the CRUD methods on the same interface.

JdbcDelegationPolicyService, EmptyDelegationPolicyService, DelegationPolicyServiceFactory,
and the DDL scripts for standard, Derby, and Oracle dialects are added to gateway-server,
following the same structure used for JdbcTrustedOidcIssuerService. The factory activates
the JDBC implementation when a KNOXIDF or KNOXIDF_ADMIN topology is deployed, and falls back
to the empty stub otherwise.

The evaluate() implementation checks policies by field: user, groups, headless exchange
flag, subject user list, resource, and per-resource scope. A resource with no defined scopes is
treated as allowing any scope. canActFor.groups support is not yet implemented; if a policy has
a non-empty groups list and the subject is not in the users list, evaluate() returns a server
error rather than silently skipping the group check.

## How was this patch tested?

Schema tests verify that derby and core sql DDL scripts create the five tables cleanly. The oracle
DDL script is validated to be present. Unit testing focuses on the a Derby in-memory
database. Tests check that NOT NULL, DEFAULT, UNIQUE, and foreign key constraints are
enforced. Default values (status='active', allow_headless_exchange=false, token_ttl_sec=null)
are verified by inserting a row without those fields. Constraint violations for missing required
fields and duplicate (actor_authority, actor_id) are confirmed.

CRUD lifecycle tests cover register, get, update (full overwrite), and delete. Update is verified
to replace child rows completely: registering with two users and updating with three different
users results in exactly three users, not five. Round-trip tests verify that resourcePolicy
entries, including resources with empty scope sets (all-scopes-allowed), are preserved across
register and get, with boundary cases for zero resources, one resource, and multiple resources,
and for resources with no scopes, one scope, and multiple scopes.

evaluate() tests cover the authorized path (actor registered, subject in users list, resource in
policy, scope in resource scopes), the all-scopes-allowed path (resource present but no scope
rows), and all denial paths: actor not registered, subject not in users list (with empty groups
list), both users and groups lists empty (deny - no open delegation), resource not in policy,
scope not in resource scopes, and headless exchange denied by policy flag. The not-yet-implemented
server error path is verified when a non-empty groups list is present and the user check fails.
If no TTL is specified, the default is used, and if specified, the policy value is used.

Factory tests confirm that the JDBC implementation is returned when KNOXIDF or KNOXIDF_ADMIN is
deployed and that the empty stub is returned otherwise.

## Integration Tests
No integration tests are added in this pull request. They will be added once the full delegation
enforcement flow is wired in a later task.

### Opt-in test suites (PR labels)
N/A

## UI changes
N/A

### CODE REVIEW

File: gateway-spi/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/PolicyCheckRequest.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  public PolicyCheckRequest(String actorAuthority, String actorId, String subjectName,
+      String requestedResource, Set<String> requestedScopes, boolean headlessExchange) {
+    this.actorAuthority = actorAuthority;
+    this.actorId = actorId;
+    this.subjectName = subjectName;
+    this.requestedResource = requestedResource;
+    this.requestedScopes = requestedScopes;
+    this.headlessExchange = headlessExchange;
+  }
```
[the remark this answers is not in the document]
Comment (lmccay) [reply gateway-spi/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/PolicyCheckRequest.java:42]: Seems fair.

File: gateway-server/src/test/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicySchemaTest.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  @BeforeClass
+  public static void setUp() throws Exception {
+    java.util.Locale.setDefault(java.util.Locale.US);
+    derbyConn = DriverManager.getConnection(DERBY_URL);
+    hsqlConn = DriverManager.getConnection(HSQL_URL, HSQL_USER, HSQL_PASSWORD);
+    // Run Derby DDL once - no IF NOT EXISTS, so run once at class level
+    runScript(derbyConn, loadSql(AbstractDataSourceFactory.DERBY_KNOXIDF_DELEGATION_POLICY_TABLES_SQL));
+    // Run standard DDL on HSQLDB
+    runScript(hsqlConn, loadSql(AbstractDataSourceFactory.KNOXIDF_DELEGATION_POLICY_TABLES_SQL));
+  }
```
[the remark this answers is not in the document]
Comment (lmccay) [reply gateway-server/src/test/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicySchemaTest.java:67]: This is understandable given the lack of oracle backend to test with. @hsheinblatt what do you think about the suggestion here?

File: gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicyDatabase.java [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+   * {@link JDBCUtils#createTableFromSQL} handles only single statements; delegation needs five.
+   * Comment stripping must happen before the split because the ASF license header contains a
+   * semicolon inside a {@code --} comment line, which would otherwise produce a spurious token.
+   */
+  private void createDelegationTablesIfNotExists(String sqlFileName) throws Exception {
+    if (!JDBCUtils.tableExists(CORE_TABLE, dataSource)) {
+      try (InputStream is = getClass().getClassLoader().getResourceAsStream(sqlFileName);
+           Connection connection = dataSource.getConnection()) {
+        final String script = IOUtils.toString(is, UTF_8);
+        final StringBuilder stripped = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (lmccay) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicyDatabase.java:149]: Fair suggestion.

File: gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicyDatabase.java [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+
+  private static final String DELETE_RESOURCES_SQL =
+      "DELETE FROM DELEGATION_POLICY_RESOURCES WHERE registration_id = ?";
+
+  DelegationPolicyDatabase(DataSource dataSource, String dbType, int listMaxTotal, int listMaxPerAuthority) throws Exception {
+    super(dataSource);
+    this.listMaxTotal = listMaxTotal;
+    this.listMaxPerAuthority = listMaxPerAuthority;
+    this.selectAllSql = SELECT_ALL_BASE_SQL + " FETCH FIRST " + (listMaxTotal + 1) + " ROWS ONLY";
+    this.selectAllFilteredSql = SELECT_ALL_BASE_SQL + " WHERE actor_authority = ? FETCH FIRST " + (listMaxPerAuthority + 1) + " ROWS ONLY";
```
[the remark this answers is not in the document]
Comment (lmccay) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicyDatabase.java:129]: Interesting suggestion. @hsheinblatt - thoughts? Seems like any test scenarios would benefit from that being more deterministic.

File: gateway-spi/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicy.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    this.createdAt = createdAt;
+    this.updatedAt = updatedAt;
+    this.allowHeadlessExchange = allowHeadlessExchange;
+    this.canActForUsers = Collections.unmodifiableSet(new HashSet<>(canActForUsers));
+    this.canActForGroups = Collections.unmodifiableSet(new HashSet<>(canActForGroups));
+    Map<String, Set<String>> copy = new HashMap<>();
+    for (Map.Entry<String, Set<String>> entry : resourcePolicy.entrySet()) {
+      copy.put(entry.getKey(), Collections.unmodifiableSet(new HashSet<>(entry.getValue())));
+    }
+    this.resourcePolicy = Collections.unmodifiableMap(copy);
```
[the remark this answers is not in the document]
Comment (lmccay) [reply gateway-spi/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicy.java:69]: Seems fair enough.

File: gateway-server/src/test/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    EasyMock.expect(gws.getService(ServiceType.TOPOLOGY_SERVICE)).andReturn(topologyService).anyTimes();
+    EasyMock.expect(gws.getService(ServiceType.ALIAS_SERVICE)).andReturn(aliasService).anyTimes();
+    EasyMock.replay(gws);
+
+    final GatewayConfigImpl config = EasyMock.createNiceMock(GatewayConfigImpl.class);
+    EasyMock.expect(config.getDatabaseType()).andReturn(DatabaseType.DERBY.type()).anyTimes();
+    EasyMock.expect(config.getDatabaseName())
+        .andReturn("memory:" + getClass().getSimpleName() + "_knoxidf").anyTimes();
+    EasyMock.expect(config.getDelegationServiceTokenTtlSec()).andReturn(3600).anyTimes();
+    EasyMock.replay(config);
```
[the remark this answers is not in the document]
Comment (lmccay) [reply gateway-server/src/test/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactoryTest.java:134]: Seems fair.

File: gateway-server/src/main/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactory.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+  private static final GatewayMessages LOG = MessagesFactory.get(GatewayMessages.class);
+  private static final String DEFAULT_IMPLEMENTATION = EmptyDelegationPolicyService.class.getName();
+
+  @Override
+  protected Service createService(GatewayServices gatewayServices, ServiceType serviceType,
+      GatewayConfig gatewayConfig, Map<String, String> options, String implementation)
+      throws ServiceLifecycleException {
+
+    String implementationToUse = implementation;
```
Comment (lmccay) [inline gateway-server/src/main/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactory.java:43]: I wonder whether we need an empty implementation. We should be able to configure whether the delegation service should be added or not. I see the check for isKnoxIdfEnabledInAnyTopology and note the intent there. I'm curious of the timing of that and how it is actually known at the time of gateway services being initiatlized.

Comment (lmccay) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactory.java:43]: In general, I don't like the hard coding of classnames rather than usingthe service loader mechanism but I see that this within the factory implementation where the implementation is being provided. I need to dig into that more.

**Review by lmccay — COMMENTED** [review]
Made a couple comments and raised a concern about the timing of gateway services starting and knowing whether topologies contain knoxidf yet at that time.

Otherwise, I will leave the schema review to @smolnar82 or file a followup jira to track a review for it.

File: gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicyDatabase.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (hsheinblatt) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicyDatabase.java:129]: My opinion is that we should not add an order by. That would have performance implications for every query in normal operation. The case in question is already an error case: we've hit the scalability limit of the database and it's no longer working correctly. Clients have a flag so they can log the condition, but the administrator has to fix it. More resources can be assigned and the limit raised, rows could be cleaned up to free up capacity, but the system isn't working anymore. It seemed best to truncate results with a flag rather than throw an exception, but the client has no graceful recovery option here. There's no way to retrieve the missing rows even if there were an order by, and the data returned is incorrect because it is incomplete either way. The design could be changed to use paging instead, but would similarly still have capacity limits that could still be reached and cause error conditions.

File: gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicyDatabase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (hsheinblatt) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicyDatabase.java:149]: This one is really hard to test, and a bit unlikely to occur. This is a private method a few layers down in init. In order to get to this error path, we'd need a databasetype that succeeded lookup but was missing a valid sql resource. I can add this protection, but then I'll skip the unit test for it. Ideally one of the layers leading to it should fail at build time not runtime if there is a problem.

File: gateway-server/src/test/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicySchemaTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hsheinblatt) [reply gateway-server/src/test/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicySchemaTest.java:67]: This seems reasonable, added. There are no other similar tests for oracle. The unit tests don't show the sql exists, it's parsable, or it works. We'd need an integration test to validate oracle integration.

File: gateway-server/src/main/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactory.java (hunk 7 above)
Comment (hsheinblatt) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactory.java:43]: For the empty implementation, I was following the pattern from FederatedIdentityService, so Sandor might have a better understanding of the intent. My understanding is that the empty service is a soft config failure.

On a config failure, you could say, that's not allowed, fail knox startup, everything stops with a clear config exception. Other services do that, like DefaultTokenAuthorityService.  The pattern in the FederatedIdentityService, also used in TrustedOidcIssuer, is that if there is some failure,  it doesn't block startup, instead it loads the empty service and that will cause runtime errors if those paths are hit.

One example failure case is you've configured the delegation service enabled flag to allow policy-driven delegation token exchanges, but you haven't configured the delegation service. Arguably this should just fail startup -- you want to use the delegation service but it won't work. But possibly there are other topologies that you'd want to run and not block for this.

Depending on your deployment model, failing startup is cleaner -- for example in k8s, you'd ideally get a crash loop backoff and not deploy the new version until you fixed the config. But in other deployment models, it may cause a real problem. With this implementation, knox starts up, but then if you try a delegation token exchange, they'll all fail with a configuration error.

It also fails over to an empty config on other init failures, so if the config is incorrect or there's a runtime problem initializing, knox will still start up but with the empty service. We could make a distinction on what causes a full failure and what fails over to empty based on the type of configuration problem or runtime issue. 

The FederatedIdentity service is different: we may not need it at init time.  Since the delegation service is to have an explicit enablement flag, we can fail startup deterministically if we're comfortable failing startup.

For now I suggest we keep this failure pattern, and follow up to further justify this pattern or rethink it across the three services.

Comment (hsheinblatt) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactory.java:43]: For the conditional loading of the delegation policy service based on the configuration for either knoxidf or knoxadmin using isKnoxIdfEnabledInAnyTopology(), this has some issues. First, the primary check here is through the topology service, which is loaded after the delegation service so in this case will always fail through to the isKnoxIdfEnabledOnDisk() fallback. But even if we reordered the loading so it was initialized, it doesn't provide much value: checking the topology xml files should be correct anyway. So we could just remove that initial check and fix it to use the fallback reading of the xml files.

It's possible to dynamically load topologies from new files later. If at init time, no topologies used knoxidf or knoxadmin, so the delegation service didn't initialize, but then later a new topology was loaded that did include knoxidf or knoxadmin, the topology load wouldn't reload the gateway services, and so the delegation policy service wouldn't be reloaded for them, and they'd fail if they needed it. There are multiple paths here that cause dynamic loading.  The simplified descriptor feature seems unlikely to apply to knoxidf. But API calls or direct file manipulation are possibilities.

Arguably if you want to later add a knoxidf or knoxadmin service topology with delegation, you'd want to configure the delegation service storage, and if you hadn't already configured it, you'd need to add that configuration, and so you'd need to restart knox for the gateway service storage update. So you could have known that the dynamic loading shouldn't work.

Given this is in the abstract base class, and it works similarly to the other knoxidf services, I suggest leaving it for now. It is convenient for the standard case: you want to use knoxidf or knoxadmin, and you need the service, and you don't want to have to add boilerplate. For follow on, we can consider skipping the topology service check or requiring a specific configuration to load the delegation service.

Comment (hsheinblatt) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactory.java:43]: On the use of the factory pattern rather than the service loader: the piece of the service I could see as pluggable is the storage layer, but there's no precedent I could find in the code base for using a pluggable storage layer for the RDBMS. They all use a factory pattern and manage the SQL engine to use based on internal config. The LDAP service uses a pluggable backend, which makes sense as that's standard. Arguably knox could have done that for RDBMS backends as well, but that is not the established pattern.

The current implementation has decision logic, which also requires a factory, but as noted above, all those could be removed. But even if they were, I'd think we'd just remove the factory and directly create the jdbc service, still not needing a plugin.

File: gateway-server/src/test/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactoryTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (hsheinblatt) [reply gateway-server/src/test/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactoryTest.java:134]: added

File: gateway-spi/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicy.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (hsheinblatt) [reply gateway-spi/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicy.java:69]: Added

File: gateway-spi/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/PolicyCheckRequest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (hsheinblatt) [reply gateway-spi/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/PolicyCheckRequest.java:42]: Added

**Comment by hsheinblatt** [comment]
I've updated the PR with fixes and added comments in response to questions.

File: gateway-server/src/main/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactory.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    if (isEmptyDefaultImplementation(implementationToUse) && isKnoxIdfEnabledInAnyTopology(gatewayServices, gatewayConfig)) {
+      implementationToUse = JdbcDelegationPolicyService.class.getName();
```
Comment (smolnar82) [inline gateway-server/src/main/java/org/apache/knox/gateway/services/factory/DelegationPolicyServiceFactory.java:45]: I'd follow what we already have in `TrustedOidcIssuerServiceFactory` and `FederatedIdentityServiceFactory`: use JdbcDelegationPolicyService only when we have an external DB configured. Otherwise, use DerbyDBDelegationPolicyService (needs to be added here) that uses our embedded DerbyDB and is useful for testing locally.

File: gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicyDatabase.java [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+    final DatabaseType databaseType = DatabaseType.fromString(dbType);
+    createDelegationTablesIfNotExists(databaseType.delegationPolicyTablesSql());
```
Comment (smolnar82) [inline gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/DelegationPolicyDatabase.java:131]: Why not reuse KnoxDatabase#createTableIfNotExists here?
Ohh, because these .sql files bundle five `CREATE TABLE` statements and `JDBCUtils.createTableFromSQL` runs the file as a single `Statement.execute()`, which Derby rejects.

Rather than a private splitter here (also duplicated in `DelegationPolicySchemaTest.runScript`), could we push this into the shared layer? Two small additions:
- a JDBCUtils.createTableFromSQL(String sql, DataSource) overload that executes one already-loaded statement, and
- a KnoxDatabase.createTablesFromSQL(...) that splits the script and calls it per statement.

Another option is to split that "all in one" SQL file into several create table SQL files, which fits into what we have now and adds less complexity (not to mentioned, it's well tested).

I can see one more drawback with your approach: you only check if the `DELEGATION_POLICIES` has been created already. What happens, if that condition is true, but the previous attempt of creating ALL delegation policy tables failed in the middle. The next time, this check will be true, and we find ourselves in a place where some of the tables are missing.

All of the above points the same way for me: splitting the all-in-one script into per-table files (as we already do for the other KnoxToken/KnoxIDF tables) removes the fragile parser, the test duplication, and the partial-creation guard weakness in one move, with no new shared machinery. That'd be my preference.

**Review by smolnar82 — CHANGES_REQUESTED** [review]

File: gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/JdbcDelegationPolicyService.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+    // Step 6: effective TTL — use the policy value if set, otherwise fall back to the configured default
+    final int effectiveTtlSec = policy.getTokenTtlSec() != null
+        ? policy.getTokenTtlSec()
+        : configuredKnoxTokenTtlSec;
+
+    // Step 7: group check (LDAP lookup — slowest, deferred past cheap checks)
+    if (!userCheckPassed) {
+      if (!policy.getCanActForGroups().isEmpty()) {
+        throw new UnsupportedOperationException("canActFor.groups evaluation not yet implemented");
```
Comment (smolnar82) [inline gateway-server/src/main/java/org/apache/knox/gateway/services/knoxidf/delegation/JdbcDelegationPolicyService.java:192]: Claude's comment on this:
```
evaluate() throws on a policy shape that register()/update() persists (there are even tests storing canActForGroups), 
so this fires on a valid, stored policy — and per the design (KNOX-3426 §6/§8) groups are a first-class MVP path, not an impossible input. 
Once wired to the exchange handler this surfaces as 500 instead of the required access_denied.

Either reject non-empty canActForGroups at registration (can't persist what you can't evaluate), or at least fail closed here — return deny("subject_not_allowed") — and defer the real LDAP check to a follow-up.
```

**Comment by smolnar82** [comment]
Let me merge this PR and submit the 2 DB-related changes as a separate one.
